src/libAtomVM/externalterm.c

```diff
@@ -107,65 +96,61 @@ static term externalterm_to_term_internal(const void *external_term, size_t size
         // We need to allocate fragments as reading external terms from modules
         // is not accounted for by the compiler when it emits test_heap opcodes
         Heap heap;
         if (UNLIKELY(memory_init_heap(&heap, heap_usage) != MEMORY_GC_OK)) {
             return term_invalid_term();
         }
-        result = parse_external_terms(external_term_buf + 1, &eterm_size, copy, &heap, ctx->global);
+        result = parse_external_terms(external_term_buf + 1, &eterm_size, flags & ExternalTermCopy, &heap, ctx->global);
         memory_heap_append_heap(&ctx->heap, &heap);
     } else {
-        if (UNLIKELY(memory_ensure_free(ctx, heap_usage) != MEMORY_GC_OK)) {
+        if (UNLIKELY(memory_ensure_free_with_roots(ctx, heap_usage, num_roots, roots, MEMORY_CAN_SHRINK) != MEMORY_GC_OK)) {
             fprintf(stderr, "Unable to ensure %zu free words in heap\n", eterm_size);
             return term_invalid_term();
         }
-        result = parse_external_terms(external_term_buf + 1, &eterm_size, copy, &ctx->heap, ctx->global);
+        result = parse_external_terms(external_term_buf + 1, &eterm_size, flags & ExternalTermCopy, &ctx->heap, ctx->global);
     }
     *bytes_read = eterm_size + 1;
     return result;
 }
 
-term externalterm_to_term(const void *external_term, size_t size, Context *ctx, ExternalTermOpts opts)
-{
-    size_t bytes_read = 0;
-    return externalterm_to_term_internal(external_term, size, ctx, opts, &bytes_read, false);
-}
-
-term externalterm_to_term_copy(const void *external_term, size_t size, Context *ctx, ExternalTermOpts opts)
+term externalterm_to_term(const void *external_term, size_t size, Context *ctx, ExternalTermFlags flags)
 {
     size_t bytes_read = 0;
-    return externalterm_to_term_internal(external_term, size, ctx, opts, &bytes_read, true);
+    return externalterm_to_term_with_roots(external_term, size, ctx, flags, &bytes_read, 0, NULL);
 }
 
 enum ExternalTermResult externalterm_from_binary(Context *ctx, term *dst, term binary, size_t *bytes_read)
 {
     if (!term_is_binary(binary)) {
         return EXTERNAL_TERM_BAD_ARG;
     }
-    //
-    // Copy the binary data to a buffer (in case of GC)
-    //
     size_t len = term_binary_size(binary);
     const uint8_t *data = (const uint8_t *) term_binary_data(binary);
-    uint8_t *buf = malloc(len);
-    if (IS_NULL_PTR(buf)) {
-        fprintf(stderr, "Unable to allocate %zu bytes for binary buffer.\n", len);
-        return EXTERNAL_TERM_MALLOC;
+    term t;
+    if (term_is_heap_binary(binary)) {
+        //
+        // Copy the binary data to a buffer (in case of GC)
+        //
+        uint8_t *buf = malloc(len);
+        if (IS_NULL_PTR(buf)) {
+            fprintf(stderr, "Unable to allocate %zu bytes for binary buffer.\n", len);
+            return EXTERNAL_TERM_MALLOC;
+        }
+        memcpy(buf, data, len);
+        t = externalterm_to_term_with_roots(buf, len, ctx, ExternalTermCopy, bytes_read, 0, NULL);
```

Comment:
I'm not sure I'm understanding this, why aren't we adding the binary to our roots?
It might be a good idea to leave a short comment here.

Reply:
Is the comment clearer?
This PR is an improvement over previous implementation because we don't always copy buffer data. However, we still do for small (heap) binaries. I thought it better than duplicate code to separate size calculation and decoding without deallocating heap binary. Also, passed heap binary could be GC'd if it's no longer needed.